Add test cases for request.go file

labala-sravani opened the pull request:
Add test cases for panics and validation failures in request.go file. This increases the test coverage as well. This is for [Issue 164](https://github.com/gavv/httpexpect/issues/164)

gavv commented:
Nice, thanks for PR!

Some comments:

* It would be nice to avoid testing internal methods directly (like setupRedirects, encodeWebsocketRequest, setBody, etc). This internals are subject to change and the way how they are used from Request may change as well, making these tests less relevant. It would be better to do all tests via Request public interface, with rare exceptions like checking request `chain`.

* To test panics, we can use assert.Panics

* I think there is no need to hard-code specific panic messages in tests, so that we don't have to fix tests if we change messages a bit

labala-sravani commented:
Thanks for the review comments,
As per the second and third comment, I have modified the panic test cases to use assert.Panics and also removed the panic message hardcoding.

Will be checking the first comment. Currently, I have removed the internal req methods direct testing (except newRequest function). Please have a look in the mean time.

gavv commented on request_test.go:
```diff
@@ -2128,6 +2128,149 @@ func TestRequestRedirect(t *testing.T) {
+
+	t.Run("setupRedirect - redirect policy", func(t *testing.T) {
+		req := NewRequest(config, "METHOD", "/")
+		req.config.Client = nil
```
Client can't be nil, we're lucky that in this case request does not panic because of it, but it can change in future.

I guess instead, we should set Config.Client to any implementation which is not http.Client, e.g. we can use mockClient.

gavv replied:
Note: better set config.Client before creating request instead of setting private field after creating it.

gavv commented on request_test.go:
```diff
@@ -2128,6 +2128,149 @@ func TestRequestRedirect(t *testing.T) {
+	t.Run("setupRedirect - redirect policy", func(t *testing.T) {
+		req := NewRequest(config, "METHOD", "/")
+		req.config.Client = nil
+		req.redirectPolicy = FollowAllRedirects
```
Instead of setting private field directly, we can use WithRedirectPolicy

gavv commented on request_test.go:
```diff
@@ -2128,6 +2128,149 @@ func TestRequestRedirect(t *testing.T) {
+		req := NewRequest(config, "METHOD", "/")
+		req.config.Client = nil
+		req.redirectPolicy = defaultRedirectPolicy
+		req.maxRedirects = 0
```
Instead of setting private field directly, we can use WithMaxRedirects

gavv reviewed (commented):
Thanks, let me know when the rest will be ready. Here are a few minor comments.

gavv commented:
FYI: welcome to join discord chat if you prefer such format https://discord.gg/5SCPCuCWA9

gavv commented:
Please rebase on fresh `master`.

labala-sravani commented:
For now, I have fixed the minor comments. Rest, will check over the weekend. Thanks.

gavv commented on request_test.go:
```diff
@@ -2128,6 +2128,148 @@ func TestRequestRedirect(t *testing.T) {
+
+	reporter := newMockReporter(t)
+
+	t.Run("newRequest - requestFactory is nil", func(t *testing.T) {
+		config := Config{
+			RequestFactory: nil,
+			Client:         client,
+			Reporter:       reporter,
+		}
+
+		assert.Panics(t, func() { newRequest(nil, config, "", "") })
+	})
+
+	t.Run("newRequest - client is nil", func(t *testing.T) {
+		config := Config{
+			RequestFactory: factory,
+			Client:         nil,
+			Reporter:       reporter,
+		}
+
+		assert.Panics(t, func() { newRequest(nil, config, "", "") })
+	})
+
+	t.Run("newRequest - AssertionHandler is nil", func(t *testing.T) {
+		config := Config{
+			RequestFactory: factory,
+			Client:         client,
+			Reporter:       reporter,
+		}
+		config.AssertionHandler = nil
```
I guess in these tests, we need to set AssertionHandler, not Reporter. Request does not use reporter.

gavv commented:
Thanks, let me know when it's ready for review.

labala-sravani commented:
@gavv, I have fixed your comments.
And I tried to add test cases for missing one’s as well, but I couldn’t add. Let me know if you have any suggestions. Otherwise, please review the changes in the MR.

gavv commented:
Thanks!

gavv commented:
Small follow-up commit: bbea5832ac61548a3e727fac1f2ba01ab1c6854d

Changes:

* create config with bad client only in tests then need it
* use AssertionHandler instead of Reporter
* remove TestWithRetryDelay (seems that test does not make much sense)
* cosmetic fixes

gavv commented:
I've created a follow-up issue for remaining uncovered paths: #200
